src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializer.cs

```diff
@@ -348,16 +401,57 @@ public virtual object CreateResourceInstance(IEdmStructuredTypeReference structu
                     if (edmProperty == null)
                     {
                         ApplyDynamicResourceInNestedProperty(resourceInfoWrapper.NestedResourceInfo.Name, resource,
-                            structuredType, resourceWrapper, readContext);
+                            structuredType, resourceWrapper, nestedReadContext);
                     }
                     else
                     {
-                        ApplyResourceInNestedProperty(edmProperty, resource, resourceWrapper, readContext);
+                        ApplyResourceInNestedProperty(edmProperty, resource, resourceWrapper, nestedReadContext);
                     }
                 }
             }
         }
 
+        private ODataResourceSetWrapper CreateResourceSetWrapper(IEdmCollectionTypeReference edmPropertyType,
+            IList<ODataEntityReferenceLinkBase> refLinks, ODataDeserializerContext readContext)
+        {
+            ODataResourceSet resourceSet = new ODataResourceSet
+            {
+                TypeName = edmPropertyType.FullName(),
+            };
+
+            IEdmTypeReference elementType = edmPropertyType.ElementType();
+            ODataResourceSetWrapper resourceSetWrapper = new ODataResourceSetWrapper(resourceSet);
+            foreach (ODataEntityReferenceLinkBase refLinkWrapper in refLinks)
+            {
+                ODataResourceWrapper resourceWrapper = CreateResourceWrapper(elementType, refLinkWrapper, readContext);
+                resourceSetWrapper.Resources.Add(resourceWrapper);
+            }
+
+            return resourceSetWrapper;
+        }
+
+        private ODataResourceWrapper CreateResourceWrapper(IEdmTypeReference edmPropertyType, ODataEntityReferenceLinkBase refLink, ODataDeserializerContext readContext)
+        {
+            Contract.Assert(readContext != null);
+
+            ODataResource resource = new ODataResource
+            {
+                TypeName = edmPropertyType.FullName(),
+            };
+
+            resource.Properties = ODataResourceDeserializerHelpers.CreateKeyProperties(refLink.EntityReferenceLink.Url, readContext);
+
+            if (refLink.EntityReferenceLink.InstanceAnnotations != null)
```

Comment:
This check is superfluous since the `InstanceAnnotations` property cannot be null
See
1) https://github.com/OData/odata.net/blob/b37a90e32319636c368a2ce9e5ba2cf39cb65582/src/Microsoft.OData.Core/ODataEntityReferenceLink.cs#L34
2) https://github.com/OData/odata.net/blob/b37a90e32319636c368a2ce9e5ba2cf39cb65582/src/Microsoft.OData.Core/ODataAnnotatable.cs#L33